File: Nudge-Neurodivergent-task-tracker/recurring_task.py
"""
Recurring task functionality
"""

from datetime import datetime, timedelta, date
from enum import Enum
from typing import List, Set
from task_manager import Task, TaskCategory, ReminderType


class RecurrenceType(Enum):
    """Types of task recurrence"""
    DAILY = "Daily"
    WEEKLY = "Weekly" 
    WEEKDAYS_ONLY = "Weekdays Only"
    CUSTOM = "Custom Days"
    
    def __init__(self, display_name: str):
        self.display_name = display_name


class RecurringTask(Task):
    """Task that repeats on a schedule"""
    
    def __init__(self, name: str, first_due: datetime, reminder_type: ReminderType, 
                 category: TaskCategory, point_value: int, recurrence_type: RecurrenceType):
        super().__init__(name, first_due, reminder_type, category, point_value)
        self.recurrence_type = recurrence_type
        self.custom_days: Set[int] = set()  # Days of week (0=Monday, 6=Sunday)
        self.completion_count = 0
        
    def add_custom_day(self, day: int):
        """Add a custom day (0=Monday, 6=Sunday)"""
        if 0 <= day <= 6:
            self.custom_days.add(day)
    
    def schedule_next(self):
        """Schedule the next occurrence of this task"""
        if not self.completed:
            return
            
        self.completion_count += 1
        
        # Calculate next due date
        current_due = self.deadline
        next_due = None
        
        if self.recurrence_type == RecurrenceType.DAILY:
            next_due = current_due + timedelta(days=1)
        elif self.recurrence_type == RecurrenceType.WEEKLY:
            next_due = current_due + timedelta(weeks=1)
        elif self.recurrence_type == RecurrenceType.WEEKDAYS_ONLY:
            next_due = self._find_next_weekday(current_due)
        elif self.recurrence_type == RecurrenceType.CUSTOM:
            next_due = self._find_next_custom_day(current_due)
        
        if next_due:
            # Reset completion status for next occurrence
            self.completed = False
            self.completed_at = None
            self.deadline = next_due
    
    def _find_next_weekday(self, current: datetime) -> datetime:
        """Find next weekday (Monday-Friday)"""
        next_day = current + timedelta(days=1)
        while next_day.weekday() > 4:  # Skip weekends
            next_day += timedelta(days=1)
        return next_day
    
    def _find_next_custom_day(self, current: datetime) -> datetime:
        """Find next custom day"""
        if not self.custom_days:
            return current + timedelta(days=1)
        
        next_day = current + timedelta(days=1)
        for _ in range(7):  # Check up to a week ahead
            if next_day.weekday() in self.custom_days:
                return next_day
            next_day += timedelta(days=1)
        
        # If no custom day found in next week, return next occurrence of first custom day
        first_custom_day = min(self.custom_days)
        days_until = (first_custom_day - next_day.weekday()) % 7
        if days_until == 0:
            days_until = 7
        return next_day + timedelta(days=days_until)
    
    def get_next_due(self) -> datetime:
        """Get the next due date without scheduling"""
        if self.recurrence_type == RecurrenceType.DAILY:
            return self.deadline + timedelta(days=1)
        elif self.recurrence_type == RecurrenceType.WEEKLY:
            return self.deadline + timedelta(weeks=1)
        elif self.recurrence_type == RecurrenceType.WEEKDAYS_ONLY:
            return self._find_next_weekday(self.deadline)
        elif self.recurrence_type == RecurrenceType.CUSTOM:
            return self._find_next_custom_day(self.deadline)
        else:
            return self.deadline + timedelta(days=1)
    
    def __str__(self) -> str:
        """String representation including recurrence info"""
        base_str = super().__str__()
        return f"{base_str} | 🔄 {self.recurrence_type.display_name} (×{self.completion_count})"

File: Nudge-Neurodivergent-task-tracker/RecurringTask.java
import java.time.LocalDateTime;
import java.time.DayOfWeek;
import java.util.Set;
import java.util.HashSet;

/**
 * Represents a recurring task with flexible scheduling patterns
 */
public class RecurringTask extends Task {
    public enum RecurrenceType {
        DAILY("Daily"),
        WEEKLY("Weekly"),
        WEEKDAYS("Weekdays Only"),
        CUSTOM("Custom Pattern");
        
        private final String displayName;
        
        RecurrenceType(String displayName) {
            this.displayName = displayName;
        }
        
        public String getDisplayName() {
            return displayName;
        }
    }
    
    private RecurrenceType recurrenceType;
    private Set<DayOfWeek> customDays;
    private int intervalDays;
    private LocalDateTime nextDue;
    private int completionCount;
    private boolean isActive;
    
    public RecurringTask(String name, LocalDateTime firstDue, ReminderType reminderType,
                        TaskCategory category, int pointValue, RecurrenceType recurrenceType) {
        super(name, firstDue, reminderType, category, pointValue);
        this.recurrenceType = recurrenceType;
        this.customDays = new HashSet<>();
        this.intervalDays = 1;
        this.nextDue = firstDue;
        this.completionCount = 0;
        this.isActive = true;
        
        setupRecurrencePattern();
    }
    
    private void setupRecurrencePattern() {
        switch (recurrenceType) {
            case DAILY:
                intervalDays = 1;
                break;
            case WEEKLY:
                intervalDays = 7;
                break;
            case WEEKDAYS:
                customDays.add(DayOfWeek.MONDAY);
                customDays.add(DayOfWeek.TUESDAY);
                customDays.add(DayOfWeek.WEDNESDAY);
                customDays.add(DayOfWeek.THURSDAY);
                customDays.add(DayOfWeek.FRIDAY);
                break;
            case CUSTOM:
                // Will be set separately
                break;
        }
    }
    
    @Override
    public void markCompleted() {
        super.markCompleted();
        completionCount++;
        scheduleNext();
    }
    
    public void scheduleNext() {
        if (!isActive) return;
        
        switch (recurrenceType) {
            case DAILY:
            case WEEKLY:
                nextDue = nextDue.plusDays(intervalDays);
                break;
            case WEEKDAYS:
            case CUSTOM:
                nextDue = findNextCustomDay();
                break;
        }
        
        // Reset completion status for next occurrence
        setCompleted(false);
        updateDeadline(nextDue);
    }
    
    private LocalDateTime findNextCustomDay() {
        LocalDateTime candidate = nextDue.plusDays(1);
        
        while (!customDays.contains(candidate.getDayOfWeek())) {
            candidate = candidate.plusDays(1);
        }
        
        return candidate;
    }
    
    public void addCustomDay(DayOfWeek day) {
        customDays.add(day);
    }
    
    public void removeCustomDay(DayOfWeek day) {
        customDays.remove(day);
    }
    
    public void setIntervalDays(int days) {
        this.intervalDays = days;
    }
    
    public void deactivate() {
        this.isActive = false;
    }
    
    public void reactivate() {
        this.isActive = true;
    }
    
    // Getters
    public RecurrenceType getRecurrenceType() { return recurrenceType; }
    public Set<DayOfWeek> getCustomDays() { return new HashSet<>(customDays); }
    public int getIntervalDays() { return intervalDays; }
    public LocalDateTime getNextDue() { return nextDue; }
    public int getCompletionCount() { return completionCount; }
    public boolean isActive() { return isActive; }
    
    @Override
    public String toString() {
        String baseString = super.toString();
        String recurrenceInfo = " | " + recurrenceType.getDisplayName();
        if (completionCount > 0) {
            recurrenceInfo += " (completed " + completionCount + " times)";
        }
        return baseString + recurrenceInfo;
    }
    
    private void setCompleted(boolean completed) {
        // Helper method to reset completion status
        try {
            java.lang.reflect.Field field = Task.class.getDeclaredField("isCompleted");
            field.setAccessible(true);
            field.set(this, completed);
        } catch (Exception e) {
            // Fallback - this shouldn't happen in normal operation
        }
    }
    
    public void updateDeadline(LocalDateTime deadline) {
        // Helper method to update deadline for recurring tasks
        setDeadline(deadline);
    }
}